### PR TITLE
Migration to webpack 2

### PR DESCRIPTION
**Pull request by helfi92**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -48,33 +35,39 @@ const config = webpackMerge(preset, {
     fs: 'empty',
     tls: 'empty'
   },
+  entry: {
+    vendor: VENDOR_LIBS
```
Comment (eliperelman) [inline src/index.js]: What does vendor do?

File: src/index.js [hunk 2]
```diff
@@ -10,31 +10,18 @@ const webpackMerge = require('webpack-merge').smart;
...
+let VENDOR_LIBS = Object.keys(USER_CONFIG.dependencies).filter(d => !d.includes('neutrino'));
 
 preset.entry.index.unshift(require.resolve('babel-polyfill'));
 
-/**
- * Find best fit template.
- *
- * return preset template.ejs when no template found in project folder
- */
-function findTemplate() {
```
Comment (eliperelman) [inline src/index.js]: Why is this removed?

File: src/index.js [hunk 3]
```diff
@@ -83,71 +76,77 @@ const config = webpackMerge(preset, {
...
       }
     ]
   }
 });
 
 if (process.env.NODE_ENV !== 'test') {
   config.plugins.push(new webpack.optimize.CommonsChunkPlugin({
-    name: 'commons',
-    filename: 'commons.js',
+    names: ['vendor', 'manifest'],
```
Comment (eliperelman) [inline src/index.js]: What is the manifest chunk?

File: src/index.js [hunk 4]
```diff
@@ -158,13 +157,15 @@ if (process.env.NODE_ENV === 'development') {
       source: false
     }
   };
-  config.entry.index.unshift(`webpack-dev-server/client?${protocol}://${host}:${port}`);
+
+  config.entry.index.push(`webpack-dev-server/client?${protocol}://${host}:${port}/`);
```
Comment (eliperelman) [inline src/index.js]: Previously I think the webpack-dev-server entry was supported to be before the actual entry point, which was why it was unshifted. Is that not the case, or has it changed with Webpack 2?

File: src/index.js [hunk 5]
```diff
@@ -158,13 +157,15 @@ if (process.env.NODE_ENV === 'development') {
...
+  config.entry.index.push(`webpack/hot/dev-server`);
+  config.plugins.push(new webpack.HotModuleReplacementPlugin());
 } else if (process.env.NODE_ENV === 'production') {
-  config.plugins.push(new webpack.optimize.UglifyJsPlugin({
-    sourceMap: false,
-    minimize: true,
-    compress: { warnings: false }
-  }));
+  config.plugins.push(
+    new webpack.optimize.UglifyJsPlugin({ sourceMap: false, compress: { warnings: false }}),
```
Comment (eliperelman) [inline src/index.js]: Does minification work properly?

File: src/index.js (hunk 1 above)
Comment (helfi92) [inline src/index.js]: The main reason behind vendor and manifest is to reap the full benefit of browser caching. vendor libs will contains all 3rd party code and that file will probably be cached for a while. 

Regarding the manifest file, Webpack generates some webpack runtime code whenever you run webpack. If you only have vendor, then it will reside there and it will update that file every time you webpack run it. To make sure the hash for the vendor doesn't change on every run, we put the webpack run time code inside manifest. See [https://webpack.js.org/guides/code-splitting-libraries/#manifest-file](https://webpack.js.org/guides/code-splitting-libraries/#manifest-file) for a better explanation.

File: src/index.js (hunk 3 above)
Comment (helfi92) [inline src/index.js]: See comment above

File: src/index.js (hunk 4 above)
Comment (helfi92) [inline src/index.js]: Hmm, I can see from the old webpack docs that they were using `unshift` but there is no explicit mention about it. For example, this webpack 1 [tutorial](http://andrewhfarmer.com/webpack-hmr-tutorial/#2-configuring-webpackdevserver-api) puts it after the actual entry point. Also, this webpack 2 [tutorial](https://youtu.be/eWmkBNBTbMM?t=23m29s) puts it after.

File: src/index.js (hunk 2 above)
Comment (helfi92) [inline src/index.js]: At the time, I thought `PRESET_TEMPLATE_EJS` and `PROJECT_TEMPLATE` were pointing to the same place, so I changed it into a ternary. Good catch, I'll revert it.

Comment (helfi92) [inline src/index.js]: Inside the function, `PRESET_TEMPLATE` is being called but this variable doesn't exist. Did you mean `PROJECT_TEMPLATE` instead?

Comment (eliperelman) [inline src/index.js]: Check out the diff for lines 13-14. They show the variables have been changed and removed.

File: src/index.js (hunk 1 above)
Comment (eliperelman) [inline src/index.js]: Awesome thank you!

File: src/index.js [hunk 6]
```diff
@@ -10,31 +10,18 @@ const webpackMerge = require('webpack-merge').smart;
...
-const PRESET_TEMPLATE_INDEX = path.join(__dirname, 'index.html');
 const PROJECT_TEMPLATE = path.join(SRC, 'template.ejs');
+const PRESET_TEMPLATE = path.join(__dirname, 'template.ejs');
 const FILE_LOADER = require.resolve('file-loader');
 const CSS_LOADER = require.resolve('css-loader');
 const STYLE_LOADER = require.resolve('style-loader');
 const URL_LOADER = require.resolve('url-loader');
 const MODULES = path.join(__dirname, '../node_modules');
+const USER_CONFIG = require(path.join(CWD, 'package.json'));
+let VENDOR_LIBS = Object.keys(USER_CONFIG.dependencies).filter(d => !d.includes('neutrino'));
```
Comment (eliperelman) [inline src/index.js]: I think this can be `const`. Also, neutrino stuff should be dev dependencies, so the filter probably isn't needed.

File: src/index.js (hunk 4 above)
Comment (eliperelman) [inline src/index.js]: Great! As long as the order doesn't matter, I'm much happier with this way.

File: src/index.js (hunk 5 above)
Comment (helfi92) [inline src/index.js]: For a react application with HMR, those are the current stats:

Without minification:
without-minify

With minification:
with-minify
